File: packages/voltaire-zig/examples/primitives/trie/state-root.ts
/**
 * State Root Computation
 *
 * Demonstrates using a Merkle Patricia Trie to compute Ethereum state root.
 * Each account address maps to its account state (nonce, balance, storage root, code hash).
 *
 * NOTE: This example shows the intended TypeScript API. Trie is currently implemented
 * in Zig only and not yet exposed to TypeScript through FFI bindings.
 */

import { Address, Bytes32, Hash } from "@tevm/voltaire";
import type { AddressType } from "@tevm/voltaire/Address";
import type { Bytes32Type } from "@tevm/voltaire/Bytes32";
import type { HashType } from "@tevm/voltaire/Hash";

// Conceptual Trie implementation for demonstration
class Trie {
	private data = new Map<string, Uint8Array>();

	put(key: AddressType, value: Uint8Array): void {
		const hex = Address.toHex(key);
		this.data.set(hex, value);
	}

	get(key: AddressType): Uint8Array | null {
		const hex = Address.toHex(key);
		return this.data.get(hex) || null;
	}

	rootHash(): HashType | null {
		if (this.data.size === 0) return null;
		return Hash.random();
	}
}

/** Simplified account state structure */
interface AccountState {
	nonce: bigint;
	balance: bigint;
	storageRoot: Bytes32Type;
	codeHash: Bytes32Type;
}

/** Encode account state to bytes (simplified - production uses RLP) */
function encodeAccountState(state: AccountState): Uint8Array {
	const buffer = new ArrayBuffer(8 + 16 + 32 + 32);
	const view = new DataView(buffer);

	// Nonce (8 bytes)
	view.setBigUint64(0, state.nonce, false);

	// Balance (16 bytes) - simplified
	view.setBigUint64(8, state.balance & 0xffffffffffffffffn, false);
	view.setBigUint64(16, state.balance >> 64n, false);

	// Storage root (32 bytes)
	const result = new Uint8Array(buffer);
	result.set(state.storageRoot, 24);

	// Code hash (32 bytes)
	result.set(state.codeHash, 56);

	return result;
}

// Create state trie
const stateTrie = new Trie();

// Account 1: EOA with balance
const addr1 = Address(0x12n);
const account1: AccountState = {
	nonce: 5n,
	balance: 1_000_000_000_000_000_000n, // 1 ETH in wei
	storageRoot: Bytes32.zero(), // Empty storage
	codeHash: Bytes32.zero(), // No code (EOA)
};

const encoded1 = encodeAccountState(account1);
stateTrie.put(addr1, encoded1);

// Account 2: Contract with storage
const addr2 = Address(0xabn);
const account2: AccountState = {
	nonce: 1n,
	balance: 500_000_000_000_000_000n, // 0.5 ETH
	storageRoot: Bytes32.fromHex(
		"0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
	), // Has storage
	codeHash: Bytes32.fromHex(
		"0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
	), // Has code
};

const encoded2 = encodeAccountState(account2);
stateTrie.put(addr2, encoded2);

// Account 3: Another EOA
const addr3 = Address(0x34n);
const account3: AccountState = {
	nonce: 0n,
	balance: 2_500_000_000_000_000_000n, // 2.5 ETH
	storageRoot: Bytes32.zero(),
	codeHash: Bytes32.zero(),
};

const encoded3 = encodeAccountState(account3);
stateTrie.put(addr3, encoded3);

// Compute state root
const stateRoot = stateTrie.rootHash();

const account1Updated: AccountState = {
	...account1,
	nonce: 6n, // Nonce increased
};

const encoded1Updated = encodeAccountState(account1Updated);
stateTrie.put(addr1, encoded1Updated);

const newStateRoot = stateTrie.rootHash();

/**
 * Usage Pattern:
 *
 * The state trie maps account addresses (20 bytes) to RLP-encoded account state.
 * The root hash of this trie is stored in each block header, creating a
 * cryptographic commitment to the entire Ethereum world state.
 *
 * Any change to any account (balance, nonce, storage, code) changes the state root,
 * making it easy to verify that two nodes have the same state.
 */

File: packages/voltaire-zig/examples/primitives/trie/state-root.zig
//! State Root Computation
//!
//! Demonstrates using a Merkle Patricia Trie to compute Ethereum state root.
//! Each account address maps to its account state (nonce, balance, storage root, code hash).

const std = @import("std");
const primitives = @import("primitives");
const Trie = primitives.Trie;

/// Simplified account state structure
const AccountState = struct {
    nonce: u64,
    balance: u128,
    storage_root: [32]u8,
    code_hash: [32]u8,

    /// Encode account state to bytes (simplified - production uses RLP)
    pub fn encode(self: AccountState, allocator: std.mem.Allocator) ![]u8 {
        var list = std.ArrayList(u8).init(allocator);
        defer list.deinit();

        try list.writer().writeInt(u64, self.nonce, .big);
        try list.writer().writeInt(u128, self.balance, .big);
        try list.appendSlice(&self.storage_root);
        try list.appendSlice(&self.code_hash);

        return try list.toOwnedSlice();
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("=== State Root Computation ===\n\n", .{});

    // Create state trie
    var state_trie = Trie.init(allocator);
    defer state_trie.deinit();

    // Account 1: EOA with balance
    const addr1 = [_]u8{0x12} ++ [_]u8{0} ** 19; // 20-byte address
    const account1 = AccountState{
        .nonce = 5,
        .balance = 1_000_000_000_000_000_000, // 1 ETH in wei
        .storage_root = [_]u8{0} ** 32, // Empty storage
        .code_hash = [_]u8{0} ** 32, // No code (EOA)
    };
    const encoded1 = try account1.encode(allocator);
    defer allocator.free(encoded1);
    try state_trie.put(&addr1, encoded1);
    std.debug.print("Account 1: 0x{x:0>40}\n", .{std.fmt.fmtSliceHexLower(&addr1)});
    std.debug.print("  Nonce: {}\n", .{account1.nonce});
    std.debug.print("  Balance: {} wei (1 ETH)\n", .{account1.balance});

    // Account 2: Contract with storage
    const addr2 = [_]u8{0xAB} ++ [_]u8{0} ** 19;
    const account2 = AccountState{
        .nonce = 1,
        .balance = 500_000_000_000_000_000, // 0.5 ETH
        .storage_root = [_]u8{0xAA} ** 32, // Has storage
        .code_hash = [_]u8{0xBB} ** 32, // Has code
    };
    const encoded2 = try account2.encode(allocator);
    defer allocator.free(encoded2);
    try state_trie.put(&addr2, encoded2);
    std.debug.print("\nAccount 2: 0x{x:0>40}\n", .{std.fmt.fmtSliceHexLower(&addr2)});
    std.debug.print("  Nonce: {}\n", .{account2.nonce});
    std.debug.print("  Balance: {} wei (0.5 ETH)\n", .{account2.balance});

    // Account 3: Another EOA
    const addr3 = [_]u8{0x34} ++ [_]u8{0} ** 19;
    const account3 = AccountState{
        .nonce = 0,
        .balance = 2_500_000_000_000_000_000, // 2.5 ETH
        .storage_root = [_]u8{0} ** 32,
        .code_hash = [_]u8{0} ** 32,
    };
    const encoded3 = try account3.encode(allocator);
    defer allocator.free(encoded3);
    try state_trie.put(&addr3, encoded3);
    std.debug.print("\nAccount 3: 0x{x:0>40}\n", .{std.fmt.fmtSliceHexLower(&addr3)});
    std.debug.print("  Nonce: {}\n", .{account3.nonce});
    std.debug.print("  Balance: {} wei (2.5 ETH)\n", .{account3.balance});

    // Compute state root
    const state_root = state_trie.root_hash();
    std.debug.print("\n=== State Root ===\n", .{});
    std.debug.print("Root Hash: 0x{x:0>64}\n", .{std.fmt.fmtSliceHexLower(&state_root.?)});

    // Demonstrate that state root changes with state
    std.debug.print("\n=== State Mutation ===\n", .{});
    std.debug.print("Updating Account 1 nonce to 6...\n", .{});
    const account1_updated = AccountState{
        .nonce = 6, // Nonce increased
        .balance = account1.balance,
        .storage_root = account1.storage_root,
        .code_hash = account1.code_hash,
    };
    const encoded1_updated = try account1_updated.encode(allocator);
    defer allocator.free(encoded1_updated);
    try state_trie.put(&addr1, encoded1_updated);

    const new_state_root = state_trie.root_hash();
    std.debug.print("New Root Hash: 0x{x:0>64}\n", .{std.fmt.fmtSliceHexLower(&new_state_root.?)});
    std.debug.print("\n✓ Root hash changed (state is cryptographically committed)\n", .{});
}
